src/compiler/compile-props.js

```diff
@@ -342,9 +329,46 @@ export function coerceProp (prop, value) {
   return coerce(value)
 }
 
-function formatType (val) {
-  return val
-    ? val.charAt(0).toUpperCase() + val.slice(1)
+function assertType (value, type) {
+  var valid
+  var expectedType
```

Comment:
By the way, checking the `constructor` or `constructor.name` is not reliable enough?

```js
let ok = 'test'
ok.constructor === String // true
```